.github/copilot-instructions.md

```diff
@@ -0,0 +1,316 @@
+
+Runs on every push and pull request. **All jobs must pass for PR to merge.**
+
+1. **lint**: Runs `pnpm run lint` (includes prettier check)
+2. **test**: Runs `pnpm run test-ci`
+3. **typecheck**: Runs `pnpm run typecheck`
+4. **build**: Runs `pnpm run build`
+
+Each job:
+
+- Uses `ubuntu-latest` runner
+- 5-minute timeout
+- Sets `CI=true` environment variable
+- Uses `oBusk/action-pnpm-setup@v1` for Node/pnpm setup
+
+### CodeQL Scanning (.github/workflows/codeql.yml)
+
+- Runs on push to main, PRs to main, and weekly schedule
+- Analyzes JavaScript/TypeScript and GitHub Actions
+- Build mode: none (no build required for JS/TS analysis)
+
+### Dependency Review (.github/workflows/dependency-review.yml)
+
+- Runs on pull requests only
+- Blocks PRs that introduce known-vulnerable packages
+
+## Common Issues & Workarounds
+
+### Node Version Warning
+
+```
+WARN Unsupported engine: wanted: {"node":"24.x"} (current: {"node":"v20.19.6","pnpm":"10.26.0"})
+```
+
+**Safe to ignore** - the project works with Node 20.x, but CI uses Node 24.x. Tests, builds, and all commands work fine.
+
+### Clean Build
+
+If builds are failing or you see stale type errors:
+
+```bash
+rm -rf .next
+pnpm run build
+```
+
+The `.next/` directory is auto-generated and can be safely deleted.
+
+### Git Hooks
+
+Pre-commit hooks are automatically installed via `simple-git-hooks` during `pnpm install`. They run `pnpm run lint-staged` which:
+
+- Runs `eslint --fix` on staged TS/TSX/JS/JSX files
+- Runs `prettier --write` on staged MD/YAML/JSON files
+
+### Test File Pattern
+
+Tests are colocated with source files using `.test.ts` or `.test.tsx` extensions. When adding tests:
+
+- Place them next to the file they test (e.g., `utils.ts` → `utils.test.ts`)
+- Use `@testing-library/react` for component tests
+- Use `@testing-library/jest-dom` matchers (e.g., `toBeInTheDocument`)
+
+### Vercel Deployment
+
+The app is deployed to Vercel automatically. No build configuration needed - Vercel auto-detects Next.js and uses the correct build command.
+
+## Architecture Notes
+
+### API Route Structure
+
+API routes use a special `-` prefix to avoid conflicts with the catch-all `[...parts]` route:
+
+- `/api/-/versions` - Get package versions (valid endpoint)
+- `/api/example-package` - Would match the catch-all and try to treat as package name
+- See `src/app/api/-/README.md` for details
+
+### Diff Flow
+
+1. User enters package specs (e.g., `react@17.0.0...react@18.0.0`)
+2. `src/lib/destination/` canonicalizes specs and redirects if needed
+3. `src/app/[...parts]/page.tsx` renders the diff page
+4. `src/lib/npmDiff/` fetches diff using `libnpmdiff` (server-side)
+5. `src/lib/gitDiff/` parses diff into React components
+6. `react-diff-view` renders the visual diff
+
+### Server vs. Client Components
+
+- Most components are React Server Components (no `"use client"` directive)
+- Client components have `.client.tsx` extension or explicit `"use client"` directive
```

Comment:
Clarificiation: the "use client" is REQUIRED to mark components as clientside, the .client suffix is used when there is a non-client file with the same name.